47330 - FES V2 validations - Round2 - Part 3

adamf-va opened the pull request:
## Summary of changes
 
  1. Section 7.o - REOPEN Action Type - Removed
  - FES Doc: Lines showing REOPEN not supported
  - Implementation: DELETED (lines removed) - Schema handles enum constraint
  - Reason: Avoid duplicate errors with schema

  2. Section 7.r - Invalid Calendar Dates ✅ Implemented
  - FES Doc: approximateDate invalid date combinations
  - Code: Lines 588-594
  - Error: "2018-02-30 is not a valid date."
  - Validates: Feb 30, Apr 31, Feb 29 in non-leap years

  3. Section 7.t - Future Date Validation ✅ Implemented
  - FES Doc: approximateDate must be in the past
  - Code: Lines 625-630
  - Error: "The approximateDate (index) is not valid."
  - Validates: Dates cannot be in future (today is OK)

  4. Section 5.b - mailingAddress Validations ✅ Implemented
  - 5.b.ii.3: State for USA (lines 354-360) - "The state is required if the country is USA"
  - 5.b.ii.4: ZipFirstFive for USA (lines 362-368) - "The zipFirstFive is required if the country is 
  USA."
  - 5.b.iii.2: City for INTERNATIONAL (lines 336-342)
  - 5.b.iii.3: Country for INTERNATIONAL (lines 344-350)
  - 5.b.v: internationalPostalCode logic (lines 408-415, 370-377)
  - 5.b.vi: Country validation against BRD (lines 394-402)

  5. Section 5.c - changeOfAddress Validations ✅ Implemented
  - 5.c.i: TEMPORARY dates required (lines 427-444)
  - 5.c.ii: PERMANENT dates optional (lines 446-448)
  - 5.c.iii: TEMPORARY beginDate must be future (lines 288-301)
  - 5.c.iv: Date chronology (lines 291-310)
  - 5.c.v: USA address fields (lines 475-502)
  - 5.c.vi: INTERNATIONAL city/country (lines 511-529)
  - 5.c.viii: INTERNATIONAL postal code (lines 531-539)

  6. Additional Changes Not in Original PR Description:
  - BRD Mocking: Added fetch_countries_list method (lines 665-673) for local testing
  - Error Message Updates: All messages updated to match existing disability_compensation_validation.rb

  Summary of Scope Changes:

  Original PR Description (Screenshot):
  - 7.o - REOPEN validation
  - 7.q - approximateBeginDate month (1-12)
  - 7.r - approximateBeginDate day (1-31)
  - 7.s - approximateBeginDate day/year combo
  - 7.t - approximateBeginDate in past

  Actual Implementation:
  - 7.o - Removed (schema handles)
  - 7.q - Not needed (schema regex handles)
  - 7.r - ✅ Invalid date combinations only
  - 7.s - ✅ Merged into 7.r logic
  - 7.t - ✅ Future date check
  - Added: Section 5.b & 5.c address validations
  - Added: BRD mocking for testing
  - Added: Error message standardization

rockwellwindsor-va commented on modules/claims_api/app/controllers/concerns/claims_api/v2/revised_disability_compensation_validation.rb:
```diff
@@ -536,6 +539,140 @@ def validate_intl_change_postal_code!(change_of_address)
+        # Parse the date components
+        year = date_hash['year']
+        month = date_hash['month']
+        day = date_hash['day']
```
Isn't this validation handled by the schema?

adamf-va commented:
> The changing error messages will need to be rolled back and the `approximateBeginDate` only exists on `confinements` so any handling related to that will need to be altered to the actual field.
> 
> Now that you are getting a CCG token and have the ssh stuff worked out make sure to be testing these requests by sending them in to make sure.

```
{
    "data": {
        "type": "claims_api_auto_established_claim_validation",
        "attributes": {
            "status": "valid"
        }
    }
}
```

rockwellwindsor-va commented on modules/claims_api/app/controllers/concerns/claims_api/v2/revised_disability_compensation_validation.rb:
```diff
@@ -556,120 +556,40 @@ def validate_disability_action_type_reopen!(disability, index)
         collect_error(
           source: "/disabilities/#{index}/disabilityActionType",
-          title: 'Bad Request',
+          title: 'Unprocessable Entity',
```
Getting closer, but I think we can remove this `title` line and set it as a default value.  I realize some of these errors are new, but I suspect even in that case we want them to align with the current errors being returned.

rockwellwindsor-va commented on modules/claims_api/app/controllers/concerns/claims_api/v2/revised_disability_compensation_validation.rb:
```diff
@@ -570,15 +569,31 @@ def validate_approximate_date!(disability, index)
+                 else
+                   # Year only: use end of year for future check
+                   Date.new(year, 12, 31)
+                 end
```
Where does it specify to use these as the defaults? Looking at the existing check it appears we compare like values, not default to end of the year.  Does it specify to use the end of year for these validations somewhere?

Also the match for the existing check would be inclusive here, so `>=`

I believe for a year only date this is changing the validation.  The current check checks YYYY <= YYYY so the equivalent would be YYYY >= YYYY here.

So this for date `date = Date.new(2025, 12, 31)` against `Date.current` (2025-09-15) in the old validation would be true because 2025 <= 2025 would be true, but with your change to the existing behavior here that would return false.

adamf-va replied:
Just a note on the inclusion, the FES doc does specifically say `must be in the past`, so I would take that as exclusive, unless I'm missing something.

rockwellwindsor-va replied:
yep same, I'm just stating that looking in the validations currently being done, it currently does allow <=  which I believe was added to handle things like year - year, since 2025 - 2025 is probably what we want to allow but I'm not sure so I'll ask about it in stand up just to get some clarification.  

```
Everything that happens now, is happening now"
```

rockwellwindsor-va commented on modules/claims_api/app/controllers/concerns/claims_api/v2/revised_disability_compensation_validation.rb:
```diff
@@ -432,14 +436,14 @@ def validate_temporary_address_dates!(change_of_address)
           collect_error(
             source: '/changeOfAddress/dates/endDate',
-            detail: 'EndingDate is required for temporary address'
+            detail: 'Date in YYYY-MM-DD the changed address expires, if change is temporary.'
```
This reads really weird.  Not sure if some text got missed or something but it does not make much sense.

It looks like this could be a new validation we are enforcing so l would suggest to ask in the shared channel in slack for perhaps what we would want this to read as.

rockwellwindsor-va commented on modules/claims_api/app/controllers/concerns/claims_api/v2/revised_disability_compensation_validation.rb:
```diff
@@ -307,8 +309,8 @@ def validate_dates_chronological_order!(begin_date, end_date)
           source: '/changeOfAddress/dates/beginDate',
-          title: 'Invalid beginningDate',
-          detail: "BeginningDate cannot be after endingDate: #{begin_date}"
+          title: 'Invalid beginDate',
```
Clean this out

rockwellwindsor-va commented:
👀

rockwellwindsor-va commented on modules/claims_api/app/controllers/concerns/claims_api/v2/revised_disability_compensation_validation.rb:
```diff
@@ -307,8 +308,7 @@ def validate_dates_chronological_order!(begin_date, end_date)
           source: '/changeOfAddress/dates/beginDate',
-          title: 'Invalid beginningDate',
-          detail: "BeginningDate cannot be after endingDate: #{begin_date}"
+          detail: 'Begin date for the Veteran\'s new address.'
```
Is this the message you get when you test locally for this scenario against the current validations?  Looks like it would actually return
```
detail: 'beginDate is not a valid date.'
```

rockwellwindsor-va commented:
I'm not going to go through the rest of this.  There s just too much wrong and it has taken WAY too many revisions.  We need to figure out why we cannot get these messages dialed in.  Let's discuss this in stand up tomorrow.

adamf-va commented:
Local testing file for reference:
[part4_validation_tests.txt](https://github.com/user-attachments/files/22521381/part4_validation_tests.txt)

rockwellwindsor-va commented on modules/claims_api/app/controllers/concerns/claims_api/v2/revised_disability_compensation_validation.rb:
```diff
@@ -357,15 +355,15 @@ def validate_usa_mailing_address!(mailing_address)
           collect_error(
             source: '/veteranIdentification/mailingAddress/zipFirstFive',
-            detail: 'ZipFirstFive is required for USA addresses'
+            detail: 'Zip code (First 5 digits) for the Veteran\'s current mailing address. Required if country is USA.'
```
the original message here was correct

rockwellwindsor-va commented on modules/claims_api/app/controllers/concerns/claims_api/v2/revised_disability_compensation_validation.rb:
```diff
@@ -357,15 +355,15 @@ def validate_usa_mailing_address!(mailing_address)
           collect_error(
             source: '/veteranIdentification/mailingAddress/state',
-            detail: 'State is required for USA addresses'
+            detail: 'State for the Veteran\'s current mailing address. Required if country is USA.'
```
detail needs to be `The state is required if the country is USA`

rockwellwindsor-va commented on modules/claims_api/app/controllers/concerns/claims_api/v2/revised_disability_compensation_validation.rb:
```diff
@@ -411,7 +410,8 @@ def validate_international_postal_code!(mailing_address)
             source: '/veteranIdentification/mailingAddress/internationalPostalCode',
-            detail: 'InternationalPostalCode is required for non-USA addresses'
+            detail: 'International postal code for the Veteran\'s current mailing address. ' \
+                    'Do not include if country is USA.'
```
Little confused why there are two, but unless this is new (let me know if it is) I believe this should be `The internationalPostalCode should not be provided if the country is USA` (and we likely only need one of these (or is there a change I am overlooking?)

rockwellwindsor-va commented on modules/claims_api/app/controllers/concerns/claims_api/v2/revised_disability_compensation_validation.rb:
```diff
@@ -483,7 +479,7 @@ def validate_domestic_city!(change_of_address)
         collect_error(
           source: '/changeOfAddress/city',
-          detail: 'City is required'
+          detail: 'City for the Veteran\'s new address.'
```
Since this is for `DOMESTIC` I suspect we want the detail to be in the same fashion as the others `The city is required if the country is USA` but since this appears to be new can you ask in the shared channel to vieryf (or if you have leave a comment here so a reviewer knows please)

adamf-va commented:
@rockwellwindsor-va - Updated PR Summary, refactored to my previous misunderstanding on descriptions vs errors confusion, updated specs and ran another suite of local request tests. I did add a BRD mocking method for local testing I've had that consistent error on I think we discussed in our 3-way chat with Michael.
[part3_validation_tests_updated.txt](https://github.com/user-attachments/files/22540826/part3_validation_tests_updated.txt)
